src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -380,6 +406,8 @@ MultiLogicalPlanOptimize(MultiTreeRoot *multiLogicalPlan)
 	ExtendedOpNodeProperties extendedOpNodeProperties = BuildExtendedOpNodeProperties(
 		extendedOpNode);
 
+	extendedOpNodeProperties.pullUpIntermediateRows = requiresIntermediateRowPullUp;
```

Comment:
I wonder whether we should have an enum that defines where the boundary is, e.g.

```c
/* push down only the FROM .. WHERE .. clauses */
PUSH_DOWN_JOIN_TREE,

/* partially push down GROUP BY and/or aggregates, merge on the coordinator */
PUSH_DOWN_AGGREGATION_PARTIAL,

/* fully push down GROUP BY */
PUSH_DOWN_AGGREGATION_FULL,

/* fully push down window functions */
PUSH_DOWN_WINDOW_FUNCTIONS,

/* fully push down ORDER BY and/or LIMIT */
PUSH_DOWN_ORDER_BY_LIMIT
```

Otherwise this might get pretty clunky in combination with #3307.

For reference: https://jvns.ca/blog/2019/10/03/sql-queries-don-t-start-with-select/

Reply:
That'd be a good follow up PR. Would like to get #3307 merged & then hopefully get pushing down arbitrary aggregates when `GROUP BY` is on a partition key so that people at least avoid this fallback when they're grouping by something which intuitively should be pushed down